ownCloud/Import/ImportFilesController.swift

```diff
@@ -0,0 +1,154 @@
+//
+//  ImportFilesController.swift
+//  ownCloud
+//
+//  Created by Matthias Hühne on 10.07.19.
+//  Copyright © 2019 ownCloud GmbH. All rights reserved.
+//
+
+/*
+* Copyright (C) 2019, ownCloud GmbH.
+*
+* This code is covered by the GNU Public License Version 3.
+*
+* For distribution utilizing Apple mechanisms please see https://owncloud.org/contribute/iOS-license-exception/
+* You should have received a copy of this license along with this program. If not, see <http://www.gnu.org/licenses/gpl-3.0.en.html>.
+*
+*/
+
+import UIKit
+import ownCloudSDK
+
+class ImportFilesController: NSObject {
+
+	// MARK: - Instance variables
+	var url: URL
+
+	// MARK: - Init & Deinit
+	init(url: URL) {
+		self.url = url
+	}
+
+}
+
+extension ImportFilesController {
+
+	func accountOrImportUI() -> Bool {
+		let bookmarks : [OCBookmark] = OCBookmarkManager.shared.bookmarks as [OCBookmark]
+
+		if bookmarks.count > 1 {
+			let moreViewController = self.cardViewController(for: url)
+			if let wd = UIApplication.shared.delegate?.window {
+				let vc = wd!.rootViewController
+				if let navCon = vc as? UINavigationController, let topVC = navCon.visibleViewController {
+					OnMainThread {
+						topVC.present(asCard: moreViewController, animated: true)
+					}
+				} else {
+					OnMainThread {
+						vc?.present(asCard: moreViewController, animated: true)
+					}
+				}
+
+				return true
+			}
+		} else if bookmarks.count == 1, let bookmark = bookmarks.first {
+			self.importItemWithDirectoryPicker(with: url, into: bookmark)
+
+			return true
+		}
+
+		return false
+	}
+
+	func importItemWithDirectoryPicker(with url : URL, into bookmark: OCBookmark) {
+		OCCoreManager.shared.requestCore(for: bookmark, setup: { (_, _) in
```

Comment:
The `requestCore` for the `OCCore` needs to be balanced with a `returnCore` in another place.

The best place for this would be in the `placeholderCompletionHandler` passed to `core.importFileNamed(…)` - with a short delay of 1 or 2 seconds to give `OCCore` a chance to schedule the background `NSURLSessionTask` to upload the file.

Putting it in the `completionHandler` passed to `core.importFileNamed(…)`, on the other hand, would guarantee that the core stays open until the upload has finished. However, if the user doesn't have connectivity, that could keep the core open indefinitely.

Reply:
Please can you share a code example?

Reply:
Sure!

```swift
		if let progress = core?.importFileNamed(name,
							at: targetDirectory,
							from: url,
							isSecurityScoped: false,
							options: [OCCoreOption.importByCopying : true,
								  OCCoreOption.automaticConflictResolutionNameStyle : OCCoreDuplicateNameStyle.bracketed.rawValue],
							placeholderCompletionHandler: { (error, item) in
								if error != nil {
									Log.debug("Error uploading \(Log.mask(name)) to \(Log.mask(targetDirectory.path)), error: \(error?.localizedDescription ?? "" )")
								}

								// Return OCCore
								OCCoreManager.shared.returnCore(for: bookmark, completionHandler: nil)
							},
							resultHandler: { (error, _ core, _ item, _) in
…
```